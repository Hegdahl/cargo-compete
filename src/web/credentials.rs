use crate::shell::Shell;
use anyhow::Context as _;
use serde::{Deserialize, Serialize};
use std::{cell::RefCell, path::PathBuf};

pub(crate) fn cookies_path() -> anyhow::Result<PathBuf> {
    let data_local_dir =
        dirs::data_local_dir().with_context(|| "could not find the local data directory")?;

    Ok(data_local_dir.join("cargo-compete").join("cookies.jsonl"))
}

pub(crate) fn username_and_password<'a>(
    shell: &'a RefCell<&'a mut Shell>,
    username_prompt: &'static str,
    password_prompt: &'static str,
) -> impl 'a + FnMut() -> anyhow::Result<(String, String)> {
    move || -> _ {
        let mut shell = shell.borrow_mut();
        let username = shell.read_reply(username_prompt)?;
        let password = shell.read_password(password_prompt)?;
        Ok((username, password))
    }
}

pub(crate) fn dropbox_access_token() -> anyhow::Result<String> {
    let path = token_path("dropbox.json")?;

    let DropboxJson { access_token } = crate::fs::read_json(&path)
        .with_context(|| format!("First, save the access token to `{}`", path.display()))?;

    return Ok(access_token);

    #[derive(Deserialize)]
    struct DropboxJson {
        access_token: String,
    }
}

pub(crate) fn yukicoder_api_key(shell: &mut Shell) -> anyhow::Result<String> {
    let path = token_path("yukicoder.json")?;
    if path.exists() {
        crate::fs::read_json(path)
    } else {
        let api_key = shell.read_password("yukicoder API key: ")?;
        crate::fs::create_dir_all(path.parent().unwrap())?;
        crate::fs::write_json(path, &api_key)?;
        Ok(api_key)
    }
}

pub(crate) fn codeforces_api_key_and_secret(shell: &mut Shell) -> anyhow::Result<(String, String)> {
    let path = token_path("codeforces.json")?;

    let CodeforcesJson {
        api_key,
        api_secret,
    } = if path.exists() {
        crate::fs::read_json(path)?
    } else {
        let api_key = shell.read_password("Codeforces API key: ")?;
        let api_secret = shell.read_password("Codeforces API secret: ")?;
        let content = CodeforcesJson {
            api_key,
            api_secret,
        };
        crate::fs::create_dir_all(path.parent().unwrap())?;
        crate::fs::write_json(path, &content)?;
        content
    };

    return Ok((api_key, api_secret));

    #[derive(Deserialize, Serialize)]
    struct CodeforcesJson {
        api_key: String,
        api_secret: String,
    }
}

fn token_path(file_name: &str) -> anyhow::Result<PathBuf> {
    let data_local_dir =
        dirs::data_local_dir().with_context(|| "could not find the local data directory")?;

    Ok(data_local_dir
        .join("cargo-compete")
        .join("tokens")
        .join(file_name))
}
