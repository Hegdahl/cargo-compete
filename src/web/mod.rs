pub(crate) mod credentials;
pub(crate) mod retrieve_testcases;
pub(crate) mod url;

use std::time::Duration;

pub(crate) const TIMEOUT: Option<Duration> = Some(Duration::from_secs(30));
