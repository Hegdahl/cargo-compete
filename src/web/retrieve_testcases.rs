use crate::{
    project::{PackageMetadataCargoCompeteBin, TargetProblem, TargetProblemYukicoder},
    shell::Shell,
    web::credentials,
};
use cargo_metadata::Package;
use heck::KebabCase as _;
use indexmap::IndexMap;
use liquid::object;
use maplit::btreemap;
use snowchains_core::{
    testsuite::{Additional, BatchTestSuite, TestSuite},
    web::{
        Atcoder, AtcoderRetrieveFullTestCasesCredentials,
        AtcoderRetrieveSampleTestCasesCredentials, AtcoderRetrieveTestCasesTargets, Codeforces,
        CodeforcesRetrieveSampleTestCasesCredentials, CodeforcesRetrieveTestCasesTargets,
        CookieStorage, RetrieveFullTestCases, RetrieveTestCases, RetrieveTestCasesOutcome,
        RetrieveTestCasesOutcomeContest, RetrieveTestCasesOutcomeProblem,
        RetrieveTestCasesOutcomeProblemTextFiles, Yukicoder,
        YukicoderRetrieveFullTestCasesCredentials, YukicoderRetrieveTestCasesTargets,
    },
};
use std::{
    borrow::BorrowMut as _,
    cell::RefCell,
    collections::{BTreeMap, BTreeSet, HashSet},
    path::{Path, PathBuf},
};

pub(crate) fn dl_for_existing_package(
    package: &Package,
    package_metadata_bin: &mut IndexMap<String, PackageMetadataCargoCompeteBin>,
    bin_indexes: Option<&HashSet<String>>,
    full: bool,
    workspace_root: &Path,
    test_suite_path: &liquid::Template,
    shell: &mut Shell,
) -> anyhow::Result<()> {
    let mut atcoder_targets = btreemap!();
    let mut codeforces_targets = btreemap!();
    let mut yukicoder_problem_targets = btreemap!();
    let mut yukicoder_contest_targets = btreemap!();

    for (bin_index, PackageMetadataCargoCompeteBin { problem, .. }) in package_metadata_bin {
        if bin_indexes.map_or(true, |bin_indexes| bin_indexes.contains(bin_index)) {
            match problem {
                TargetProblem::Atcoder { contest, index, .. } => atcoder_targets
                    .entry(contest.clone())
                    .or_insert_with(BTreeMap::new)
                    .insert(index.clone(), bin_index.clone()),
                TargetProblem::Codeforces { contest, index, .. } => codeforces_targets
                    .entry(contest.clone())
                    .or_insert_with(BTreeMap::new)
                    .insert(index.clone(), bin_index.clone()),
                TargetProblem::Yukicoder(target) => match target {
                    TargetProblemYukicoder::Problem { no, .. } => {
                        yukicoder_problem_targets.insert(no.to_string(), bin_index.clone())
                    }
                    TargetProblemYukicoder::Contest { contest, index, .. } => {
                        yukicoder_contest_targets
                            .entry(contest.clone())
                            .or_insert_with(BTreeMap::new)
                            .insert(index.clone(), bin_index.clone())
                    }
                },
            };
        }
    }

    let mut outcomes = vec![];
    let mut urls = vec![];

    for (contest, mut problems) in atcoder_targets {
        let problem_indexes = problems.keys().cloned().collect();
        let outcome = dl_from_atcoder(&contest, Some(problem_indexes), full, shell)?;
        for RetrieveTestCasesOutcomeProblem { index, url, .. } in &outcome.problems {
            if let Some(bin_index) = problems.remove(index) {
                urls.push((bin_index, url.clone()));
            }
        }
        outcomes.push(outcome);
    }
    for (contest, mut problems) in codeforces_targets {
        let problem_indexes = problems.keys().cloned().collect();
        let outcome = dl_from_codeforces(&contest, Some(problem_indexes), shell)?;
        for RetrieveTestCasesOutcomeProblem { index, url, .. } in &outcome.problems {
            if let Some(bin_index) = problems.remove(index) {
                urls.push((bin_index, url.clone()));
            }
        }
        outcomes.push(outcome);
    }
    if !yukicoder_problem_targets.is_empty() {
        let nos = yukicoder_problem_targets.keys().cloned().collect();
        let outcome = dl_from_yukicoder(None, Some(nos), full, shell)?;
        for RetrieveTestCasesOutcomeProblem { index, url, .. } in &outcome.problems {
            if let Some(bin_index) = yukicoder_problem_targets.remove(index) {
                urls.push((bin_index, url.clone()));
            }
        }
        outcomes.push(outcome);
    }
    for (contest, mut problems) in yukicoder_contest_targets {
        let problem_indexes = problems.keys().cloned().collect();
        let outcome = dl_from_yukicoder(Some(&contest), Some(problem_indexes), full, shell)?;
        for RetrieveTestCasesOutcomeProblem { index, url, .. } in &outcome.problems {
            if let Some(bin_index) = problems.remove(index) {
                urls.push((bin_index, url.clone()));
            }
        }
        outcomes.push(outcome);
    }

    for outcome in outcomes {
        save_test_cases(workspace_root, test_suite_path, outcome, shell)?;
    }

    let mut added_url = false;
    let mut new_package_metadata_bin =
        crate::fs::read_to_string(&package.manifest_path)?.parse::<toml_edit::Document>()?;
    let bin = &mut new_package_metadata_bin["package"]["metadata"]["cargo-compete"]["bin"];

    for (bin_index, url) in urls {
        let bin_url = &mut bin[bin_index]["problem"]["url"];
        if bin_url.is_none() {
            *bin_url = toml_edit::value(url.as_str());
            added_url = true;
        }
    }

    if added_url {
        crate::fs::write(&package.manifest_path, new_package_metadata_bin.to_string())?;
        shell.status("Modified", package.manifest_path.display())?;
    }
    Ok(())
}

pub(crate) fn dl_from_atcoder(
    contest: &str,
    problems: Option<BTreeSet<String>>,
    full: bool,
    shell: &mut Shell,
) -> anyhow::Result<RetrieveTestCasesOutcome> {
    let targets = AtcoderRetrieveTestCasesTargets {
        contest: contest.to_owned(),
        problems,
    };

    let shell = RefCell::new(shell.borrow_mut());

    let credentials = AtcoderRetrieveSampleTestCasesCredentials {
        username_and_password: &mut credentials::username_and_password(
            &shell,
            "Username: ",
            "Password: ",
        ),
    };

    let full = if full {
        Some(RetrieveFullTestCases {
            credentials: AtcoderRetrieveFullTestCasesCredentials {
                dropbox_access_token: credentials::dropbox_access_token()?,
            },
        })
    } else {
        None
    };

    let cookie_storage = CookieStorage::with_jsonl(credentials::cookies_path()?)?;

    Atcoder::exec(RetrieveTestCases {
        targets,
        credentials,
        full,
        cookie_storage,
        timeout: crate::web::TIMEOUT,
        shell: &shell,
    })
}

pub(crate) fn dl_from_codeforces(
    contest: &str,
    problems: Option<BTreeSet<String>>,
    shell: &mut Shell,
) -> anyhow::Result<RetrieveTestCasesOutcome> {
    let targets = CodeforcesRetrieveTestCasesTargets {
        contest: contest.to_owned(),
        problems,
    };

    let shell = RefCell::new(shell.borrow_mut());

    let credentials = CodeforcesRetrieveSampleTestCasesCredentials {
        username_and_password: &mut credentials::username_and_password(
            &shell,
            "Username: ",
            "Password: ",
        ),
    };

    let cookie_storage = CookieStorage::with_jsonl(credentials::cookies_path()?)?;

    Codeforces::exec(RetrieveTestCases {
        targets,
        credentials,
        full: None,
        cookie_storage,
        timeout: crate::web::TIMEOUT,
        shell: &shell,
    })
}

pub(crate) fn dl_from_yukicoder(
    contest: Option<&str>,
    problems: Option<BTreeSet<String>>,
    full: bool,
    shell: &mut Shell,
) -> anyhow::Result<RetrieveTestCasesOutcome> {
    let targets = if let Some(contest) = contest {
        YukicoderRetrieveTestCasesTargets::Contest(contest.to_owned(), problems)
    } else {
        YukicoderRetrieveTestCasesTargets::ProblemNos(problems.unwrap_or_default())
    };

    let full = if full {
        Some(RetrieveFullTestCases {
            credentials: YukicoderRetrieveFullTestCasesCredentials {
                api_key: credentials::yukicoder_api_key(shell)?,
            },
        })
    } else {
        None
    };

    let shell = RefCell::new(shell.borrow_mut());

    Yukicoder::exec(RetrieveTestCases {
        targets,
        credentials: (),
        full,
        cookie_storage: (),
        timeout: crate::web::TIMEOUT,
        shell: &shell,
    })
}

pub(crate) fn save_test_cases(
    workspace_root: &Path,
    path: &liquid::Template,
    outcome: RetrieveTestCasesOutcome,
    shell: &mut Shell,
) -> anyhow::Result<Vec<PathBuf>> {
    let mut acc = vec![];

    let contest = outcome
        .contest
        .as_ref()
        .map(|RetrieveTestCasesOutcomeContest { id, .. }| &**id)
        .unwrap_or("problems");

    for snowchains_core::web::RetrieveTestCasesOutcomeProblem {
        index,
        mut test_suite,
        text_files,
        ..
    } in outcome.problems
    {
        let path = path.render(&object!({ "contest": contest, "problem": &index }))?;
        let path = Path::new(&path);
        let path = workspace_root.join(path.strip_prefix(".").unwrap_or(&path));

        acc.push(path.clone());

        let txt_path = |dir_file_name: &str, txt_file_name: &str| -> _ {
            path.with_file_name(index.to_kebab_case())
                .join(dir_file_name)
                .join(txt_file_name)
                .with_extension("txt")
        };

        for (name, RetrieveTestCasesOutcomeProblemTextFiles { r#in, out }) in &text_files {
            let in_path = txt_path("in", name);
            crate::fs::create_dir_all(in_path.parent().unwrap())?;
            crate::fs::write(in_path, &r#in)?;
            if let Some(out) = out {
                let out_path = txt_path("out", name);
                crate::fs::create_dir_all(out_path.parent().unwrap())?;
                crate::fs::write(out_path, &r#out)?;
            }
        }

        if !text_files.is_empty() {
            if let TestSuite::Batch(BatchTestSuite { cases, extend, .. }) = &mut test_suite {
                cases.clear();

                extend.push(Additional::Text {
                    path: format!("./{}", index.to_kebab_case()),
                    r#in: "/in/*.txt".to_owned(),
                    out: "/out/*.txt".to_owned(),
                    timelimit: None,
                    r#match: None,
                })
            }
        }

        crate::fs::create_dir_all(path.parent().unwrap())?;
        crate::fs::write(&path, test_suite.to_yaml_pretty())?;

        shell.status(
            "Saved",
            format!(
                "{} to {}",
                match &test_suite {
                    TestSuite::Batch(BatchTestSuite { cases, .. }) => {
                        match cases.len() + text_files.len() {
                            0 => "no test cases".to_owned(),
                            1 => "1 test case".to_owned(),
                            n => format!("{} test cases", n),
                        }
                    }
                    TestSuite::Interactive(_) => "no test cases (interactive problem)".to_owned(),
                    TestSuite::Unsubmittable => "no test cases (unsubmittable problem)".to_owned(),
                },
                if text_files.is_empty() {
                    format!("{}", path.display())
                } else {
                    format!(
                        "{}",
                        path.with_file_name(format!(
                            "{{{index}.yml, {index}/}}",
                            index = index.to_kebab_case(),
                        ))
                        .display(),
                    )
                },
            ),
        )?;
    }

    Ok(acc)
}
