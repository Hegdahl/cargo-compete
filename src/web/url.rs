use eyre::{eyre, ContextCompat as _};
use std::borrow::Cow;
use url::Url;

pub(crate) fn atcoder_contest(url: &Url) -> eyre::Result<String> {
    second_path_segment(url)
}

pub(crate) fn codeforces_contest(url: &Url) -> eyre::Result<String> {
    second_path_segment(url)
}

pub(crate) fn yukicoder_contest(url: &Url) -> eyre::Result<String> {
    second_path_segment(url)
}

fn second_path_segment(url: &Url) -> eyre::Result<String> {
    let segments = url
        .path_segments()
        .map(|ss| ss.collect::<Vec<_>>())
        .unwrap_or_default();

    let segment = segments.get(1).with_context(|| {
        format!(
            "the number of path segments is {} but the index is 1: {}",
            segments.len(),
            url,
        )
    })?;

    let decodor = || percent_encoding::percent_decode_str(segment);

    decodor()
        .decode_utf8()
        .map(Cow::into_owned)
        .map_err(|e| eyre!("{}: {}", e, decodor().decode_utf8_lossy()))
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    #[test]
    fn atcoder_contest() -> eyre::Result<()> {
        assert_eq!(
            "m-solutions2020",
            super::atcoder_contest(
                &"https://atcoder.jp/contests/m-solutions2020"
                    .parse()
                    .unwrap(),
            )?
        );
        Ok(())
    }
}
