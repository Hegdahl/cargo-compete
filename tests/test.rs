pub mod common;

use ignore::overrides::OverrideBuilder;
use insta::{assert_json_snapshot, assert_snapshot};
use liquid::object;
use once_cell::sync::Lazy;
use regex::Regex;
use snowchains_core::web::PlatformKind;
use std::io;

#[test]
fn atcoder_practice_a() -> anyhow::Result<()> {
    let (output, tree) = run(
        PlatformKind::Atcoder,
        "practice",
        "a",
        r#"---
type: Batch
timelimit: 2s
match: Lines

cases:
  - name: sample1
    in: |
      1
      2 3
      test
    out: |
      6 test
  - name: sample2
    in: |
      72
      128 256
      myonmyon
    out: |
      456 myonmyon

extend: []
"#,
        r#"use proconio::input;

fn main() {
    input! {
        a: u32,
        b: u32,
        c: u32,
        s: String,
    }

    println!("{} {}", a + b + c, s);
}
"#,
    )?;

    assert_snapshot!("atcoder_practice_a_output", output);
    assert_json_snapshot!("atcoder_practice_a_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(
    platform: PlatformKind,
    contest: &str,
    problem: &str,
    test_suite: &str,
    code: &str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |workspace_root| -> _ {
            std::fs::write(
                workspace_root.join("Cargo.toml"),
                format!(
                    r#"[workspace]
members = ["{}"]
"#,
                    contest,
                ),
            )?;

            std::fs::write(
                workspace_root.join("compete.toml"),
                liquid::ParserBuilder::with_stdlib()
                    .build()?
                    .parse(include_str!("../resources/compete.toml.liquid"))?
                    .render(&object!({
                        "template_platform": platform.to_kebab_case_str(),
                        "submit_via_binary": false,
                    }))?,
            )?;

            std::fs::create_dir_all(workspace_root.join("testcases").join(contest))?;

            std::fs::write(
                workspace_root
                    .join("testcases")
                    .join(contest)
                    .join(problem)
                    .with_extension("yml"),
                test_suite,
            )?;

            std::fs::create_dir_all(workspace_root.join(contest).join("src").join("bin"))?;

            std::fs::write(
                workspace_root.join(contest).join("Cargo.toml"),
                format!(
                    r#"[package]
name = "problems"
version = "0.1.0"
edition = "2018"

[package.metadata.cargo-compete.bin]
{problem} = {{ name = "{contest}-{problem}", problem = {{ platform = "{platform}", contest = "{contest}", index = "{problem}" }} }}

[[bin]]
name = "{contest}-{problem}"
path = "src/bin/{problem}.rs"

[dependencies]
proconio = "0.3.6"
"#,
                    contest = contest,
                    problem = problem,
                    platform = platform.to_kebab_case_str(),
                ),
            )?;

            std::fs::write(
                workspace_root
                    .join(contest)
                    .join("src")
                    .join("bin")
                    .join(problem)
                    .with_extension("rs"),
                code,
            )?;
            Ok(())
        },
        io::empty(),
        &[
            "",
            "compete",
            "t",
            problem,
            "--manifest-path",
            &format!("./{}/Cargo.toml", contest),
        ],
        |_, output| {
            macro_rules! lazy_regex(($regex:literal) => (Lazy::new(|| Regex::new($regex).unwrap())));

            static RUNNING: Lazy<Regex> = lazy_regex!("^     Running `[^`]+`");
            static ACCEPTED: Lazy<Regex> = lazy_regex!(r"Accepted \([0-9]+ ms\)");

            let output = RUNNING.replace(&output, "     Running {{ command }}");
            let output = ACCEPTED.replace_all(&output, "Accepted ({{ elapsed }}) ms)");
            output.into_owned()
        },
        |workspace_root| {
            OverrideBuilder::new(workspace_root)
                .add("!/target")?
                .build()
        },
    )
}
