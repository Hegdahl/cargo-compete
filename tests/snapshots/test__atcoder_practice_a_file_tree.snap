---
source: tests/test.rs
expression: tree
---
{
  "Cargo.lock": "..",
  "Cargo.toml": "[workspace]\nmembers = [\"practice\"]\n",
  "compete.toml": "# How to manage new workspace members (\"include\" | \"exclude\" | \"focus\")\n#\n# - `skip`:    Does not modify `[workspace]`\n# - `include`: Adds the package to `workspace.members`\n# - `exclude`: Adds the package to `workspace.exclude` and create a symlink to the `compete.toml`\n# - `focus`:   Adds the package to `workspace.members` and remove the other from both of `workspace.{members, exclude}`\nnew-workspace-member = \"include\"\n\n# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `problem`:      Problem index (e.g. \"a\", \"b\")\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"./testcases/{{ contest }}/{{ problem | kebabcase }}.yml\"\n#test-suite = \"{{ manifest_dir }}/testcases/{{ problem | kebabcase }}.yml\"\n\n# Open files with the command (`jq` command)\n#\n# VSCode:\n#open = '[\"code\"] + (.paths | map([.src, .test_suite]) | flatten) + [\"-a\", ([(.git_workdir | select(.)), .manifest_dir] | first)]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[template]\nplatform = \"atcoder\"\nmanifest = \"./cargo-compete-template/Cargo.toml\"\nsrc = \"./cargo-compete-template/src/main.rs\"\n\n#[submit-via-binary]\n#target = \"x86_64-unknown-linux-musl\"\n##cross = \"cross\"\n#strip = \"strip\"\n##upx = \"upx\"\n",
  "practice": {
    "Cargo.toml": "[package]\nname = \"problems\"\nversion = \"0.1.0\"\nedition = \"2018\"\n\n[package.metadata.cargo-compete.bin]\na = { name = \"practice-a\", problem = { platform = \"atcoder\", contest = \"practice\", index = \"a\" } }\n\n[[bin]]\nname = \"practice-a\"\npath = \"src/bin/a.rs\"\n\n[dependencies]\nproconio = \"0.3.6\"\n",
    "src": {
      "bin": {
        "a.rs": "use proconio::input;\n\nfn main() {\n    input! {\n        a: u32,\n        b: u32,\n        c: u32,\n        s: String,\n    }\n\n    println!(\"{} {}\", a + b + c, s);\n}\n"
      }
    }
  },
  "testcases": {
    "practice": {
      "a.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      1\n      2 3\n      test\n    out: |\n      6 test\n  - name: sample2\n    in: |\n      72\n      128 256\n      myonmyon\n    out: |\n      456 myonmyon\n\nextend: []\n"
    }
  }
}
